### PR TITLE
dev/core#890 - Multiple line item shown on view contribution if parti…

### PR DESCRIPTION
**Pull request by jitendrapurohit**
…cipant is transferred to another contact

Overview
----------------------------------------
Multiple line item shown on view contribution if participant is transferred to another contact

Before
----------------------------------------
To replicate -

- Create a priceset for an event with `Text / Numeric Quantity` field type. Amount = $10.
- Register contact A to the event using the above price option.
- Transfer the event participation to another contact B.
- View contribution on the main contact, it displays double line items for the participation. 

image

I think the older participant status is set to "transferred" so it should remove the contribution link from the line item table?

After
----------------------------------------

Line item is displayed correctly.

image

Comments
----------------------------------------
Gitlab - https://lab.civicrm.org/dev/core/issues/890

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit can we figure out how to get a unit test on this?

FYI I have been thinking maybe it makes sense to have a 'move' api action eg. Participant.move to move a participant record from one contact to another, contribution would also be good here. BUt even a form submit test would give me more comfort

**Comment by jitendrapurohit** [comment]
@eileenmcnaughton Added unit test and separated `transferParticipantRegistration()` to process the transfer fucntionality 👍

File: CRM/Event/Form/SelfSvcTransfer.php [hunk 1]
```diff
@@ -363,54 +386,44 @@ public function postProcess() {
...
-    $params = [1 => [$this->_from_participant_id, 'Integer']];
+    $params = [1 => [$fromParticipantID, 'Integer']];
     $dao = CRM_Core_DAO::executeQuery($query, $params);
     //copy line items to new participant
-    $line_items = CRM_Price_BAO_LineItem::getLineItems($this->_from_participant_id);
-    foreach ($line_items as $item) {
+    $line_items = CRM_Price_BAO_LineItem::getLineItems($fromParticipantID);
+    foreach ($line_items as $id => $item) {
+      //Remove contribution id from older participant line item.
+      CRM_Core_DAO::singleValueQuery("UPDATE civicrm_line_item SET contribution_id = NULL WHERE id = %1", [1 => [$id, 'Integer']]);
```
Comment (jitendrapurohit) [inline CRM/Event/Form/SelfSvcTransfer.php]: this is the only change needed for this ticket to resolve. Rest is the separation of code into a single function to pass the unit test assertion.

Comment (JoeMurray) [reply CRM/Event/Form/SelfSvcTransfer.php]: The code should not be changing financial stuff just through what appears in the UI by changing the line item table, UNLESS that is the only error. I haven't looked carefully at this, but can you confirm that the financial entries in civicrm_financial_item, civicrm_financial_trxn and civicrm_entity_financial_trxn are being update correctly. 

IIRC the registration is supposed to move to a new contact, but the contribution information is supposed to be unaffected.

If the contribution is being transferred as well as the participant registration, or if it is being cancelled, then there should be entries for the first contribution that reverse out the effect of the relevant line item, and a new contribution and line item should be added to the contact receiving the registration. But I don't think this is the case. The $10 contribution should be left alone, unchanged.

**Comment by eileenmcnaughton** [comment]
We discussed this at the sprint (as part of a general discussion) & we feel like this is stalled on concept approval - ie. not quite sure we've agreed how it should look once fixed - I'm closing for now as it's not reviewable at the moment but further discussion can be done here or in gitlab until the outcome is agreed

**Comment by lcdservices** [comment]
FWIW, I agree with @JoeMurray -- transferring a participant should not have any financial impact. The contribution and associated line items should remain with the original contact who incurred the transaction.

On a side note, I have a move participant extension: https://github.com/lcdservices/biz.lcdservices.moveparticipant

It handles things a bit differently, and doesn't have a frontend component (where end user participants can transfer the registration). All it basically does is assign the participant record to a different contact and then creates an activity record to record the move.

**Comment by eileenmcnaughton** [comment]
FWIW 'vision' is that 'move' should be another action that is available on any BAO that has a contact_id field

**Comment by jitendrapurohit** [comment]
Hi All, 

relooking into this problem. The current status of db values when a participant is transferred to another contact is -

- Duplicate line item is created for the new participant(seems correct).
- The line item row inserted holds the same contribution id. So now, 2 line items are present for a single contribution id. This is the reason for duplicate rows displayed in the view contribution form. See screenshot attached in the description.
- No financial entry is made in `civicrm_financial_item` table.


Proposed Fix -

Solution 1:

- NULLify the contribution id from the previous line item (as done in this PR).
- Modify the corresponding financial item row in `civicrm_financial_item` table so it points to the new line item id created for the new participant.

Solution 2:
- Let there be 2 line items for a single contribution.
- Modify the code used to display the line items in "View contribution" form so that it loads only a single line item.

Thoughts @eileen @lcdservices @JoeMurray

**Comment by lcdservices** [comment]
Solution 1 makes more sense to me, and is more straightforward. The old participant record is only kept for the purpose of historical reference. It should be detached from the contribution/payment details and the new (copied) participant record should now be attached to it.

**Comment by eileenmcnaughton** [comment]
Sorry - I'm confused - we should only be keeping valid line items - not setting contribution_id to NULL. It's fine to update the entity_id on the line item to a different participant record but I think you are suggesting 'orphan' line items?

**Comment by eileenmcnaughton** [comment]
In other words

 Duplicate line item is created for the new participant(seems correct).

doesn't seem correct to me

**Comment by jitendrapurohit** [comment]
Hmm, ok, I think that provides another view for solution 3

>Duplicate line item is created for the new participant(seems correct).
>doesn't seem correct to me

The reason why it looked correct to me is as we create a different participant record for the transferred registration, a new line item is needed to display on the view participant page.

I think we can proceed by duplicating the line item for the new participant, but not copying the contribution id on it. The line item is meant for the new participant id, and should not hold the contribution id of the original contact. Makes sense?

**Comment by eileenmcnaughton** [comment]
Line items need to accurately reflect what has been sold - it's OK to change the participant_id (entity_id) to move it - but there should not be 2 line items for one participant record

**Comment by jitendrapurohit** [comment]
> but there should not be 2 line items for one participant record

After the transfer is done, there are 2 participant records created(not 1). If an update is done to the entity id in the line item, the original participant record won't be attached to any line item and will result in missing information on the view participant page.

**Comment by eileenmcnaughton** [comment]
Hmm - so we keep a copy of the record that used to exist with a status of transferred. I think it's correct for it NOT to have a line item - perhaps we need to address the View screen & make it not try to show the old record? I will have to test to see what it looks like to better understand

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit if we MOVE the line item (update the participant id) rather than duplicate it the selection no longer appears on the transferred registration - but there is a link to the new contact. This seems good to me - it's been transferred so it makes sense to have less information. 

Screen Shot 2020-02-17 at 6 01 13 PM

**Comment by jitendrapurohit** [comment]
Hmm, this and discussion at https://lab.civicrm.org/dev/core/issues/184 seems to clear the plot holes in the implementation of this feature. I'll soon be creating a new PR(or update this one) with the required change. Thanks @eileenmcnaughton @lcdservices 👍

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit so this is the sort of thing I'm imagining

--- a/CRM/Event/Form/SelfSvcTransfer.php
+++ b/CRM/Event/Form/SelfSvcTransfer.php
@@ -363,15 +363,10 @@ class CRM_Event_Form_SelfSvcTransfer extends CRM_Core_Form {
     //now update registered_by_id
     $query = "UPDATE civicrm_participant cp SET cp.registered_by_id = %1 WHERE  cp.id = ({$participant->id})";
     $params = [1 => [$this->_from_participant_id, 'Integer']];
-    $dao = CRM_Core_DAO::executeQuery($query, $params);
-    //copy line items to new participant
-    $line_items = CRM_Price_BAO_LineItem::getLineItems($this->_from_participant_id);
-    foreach ($line_items as $item) {
-      $item['entity_id'] = $participant->id;
-      $item['id'] = NULL;
-      $item['entity_table'] = "civicrm_participant";
-      $new_item = CRM_Price_BAO_LineItem::create($item);
-    }
+    CRM_Core_DAO::executeQuery($query, $params);
+    //Move line items to new participant
+    $line_items = civicrm_api3('LineItem', 'get', ['participant_id' =>  $this->_from_participant_id, 'api.LineItem.create' => ['participant_id' => $participant->id]]);

**Comment by jitendrapurohit** [comment]
Ah, seems we're going against the line item rules https://github.com/civicrm/civicrm-core/blob/master/CRM/Price/BAO/LineItem.php#L45-L50. Modifying `entity_id` is not allowed in update mode. But I think transfer should be an exception. Might need to update this part too with a custom param, eg 'transferLineItem` 🤷‍♂

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit I've discovered another way  that line  items with NULL contribution_id can be created - so setting contribution_id to NULL is (sadly) OK

So your solution 1 above is acceptable

**Comment by eileenmcnaughton** [comment]
(I think deleting the line items is also acceptable but looks like you might be fighting the BAO to do it)

**Comment by jitendrapurohit** [comment]
>So your solution 1 above is acceptable

That is more likely the changes done here. I don't see an option to reopen this PR so have created another one https://github.com/civicrm/civicrm-core/pull/16956 with solution 1 changes from https://github.com/civicrm/civicrm-core/pull/14104#issuecomment-586675243. It also includes the update done to financial item entries.

**Pull request closed without merging**